include/iterlib/Item.h

```diff
@@ -0,0 +1,229 @@
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  void sync() {
+    // This is a heuristic. The actual sync depends on what was
+    // modified. Use syncIdTs/syncMap directly depending on the
+    // context.
+    if (id_ == 0) {
+      syncIdTs();
+    } else {
+      syncMap();
+    }
+  }
+
+  const dynamic& value() const {
+    return *this;
+  }
+
+  bool operator<(const ItemOptimized& other) const {
+    return std::tie(value(), ts_, id_)
+           < std::tie(other.value(), other.ts_, other.id_);
+  }
+
+  bool operator==(const ItemOptimized& other) const {
+    return std::tie(value(), id_, ts_)
+           == std::tie(other.value(), other.id_, other.ts_);
+  }
+
+  // Weaker form of operator== which ignores types.
+  // May relax the equality criteria further based on use cases.
+  bool equals(const ItemOptimized& other) const {
+    if(std::tie(id_, ts_) != std::tie(other.id_, other.ts_)) {
+      return false;
+    }
+    if (value().empty() && other.value().empty()) {
+      return true;
+    }
+    return value() == other.value();
+  }
+
+  // Used by groupby. Note that id/ts are not populated
+  ItemOptimized& operator=(const dynamic& other) {
+    dynamic::operator=(other);
+    return *this;
+  }
+
+  ItemOptimized& operator=(dynamic&& other) {
+    dynamic::operator=(other);
+    return *this;
+  }
+
+  bool isAdditionalField(const std::string& key) const {
+    return key == kIdKey || key == kTimeKey;
+  }
+
+  int64_t getAdditionalField(const std::string& key) const {
+    if (key == kIdKey) {
+      return id_;
+    } else if (key == kTimeKey) {
+      return ts_;
+    }
+    throw std::logic_error("not an additional field");
+  }
+
+  std::string toString() const {
+    return folly::sformat("[{0}, {1}, {2}]", ts(), id(), dynamic::toString());
+  }
+
+protected:
+  id_t id_;
+  int64_t ts_;
+};
+
+// lessThanOrEqual is a 3-value function:
+//
+// 1  => less than
+// =  => equal
+// -1 => not equal or less than. either uncomparable or greater than
+// Notice that this signature is different from std:: compare (two value)
+// and also different from C-style compare (greater/equal/less) which assumes
+// total ordering. This function signature is usable by comparators, and also
+// the upcoming join operator (which uses useTsIdBackup = false)
+//
+// The function performs a projection as well as comparison and deals with
+// corner cases involving some columns not being present in v1 or v2
+int lessThanOrEqual(const Item& v1,
+            const Item& v2,
+            const std::vector<std::string>& columns,
+            const std::vector<bool>& isColumnDescending,
+            bool useTsIdBackup);
+
+int lessThanOrEqual(const Item& v1,
+            const Item& v2,
+            const std::vector<std::string>& columns,
+            bool useTsIdBackup);
+
+inline std::ostream& operator<<(std::ostream& os, const Item& row) {
```

Comment:
Make this non-inline and move this out of the header. Then the header will need to include only `<iosfwd>`, not `<ostream>` or `<iostream>`.